app/models/user.rb

```diff
@@ -173,14 +179,34 @@ def api_access_token_expired?
     self.api_access_token.nil? ? true : Time.at(self.api_access_token[:expires_at]) < Time.now.in_time_zone(self.get_token_timezone(:api_access_token))
   end
 
+  # check if an API access token has 'timed out' due to no usage in the last 30 minutes
+  def api_access_token_timed_out?
+    if self.api_access_token.nil?
+      true
+    else
+      # if token has been used in last 30 min (User.timeout_in), then timestamp + timeout_in > now
+      # therefore, make sure last_access + timeout_in is in the future
+      last_access = self.api_access_token[:last_access_at]
+      last_access.present? ? (last_access + self.timeout_in) < Time.now.in_time_zone(self.get_token_timezone(:api_access_token, :last_access_at)) : true
+    end
+  end
+
+  # refresh API token last_access_at, if token is present
+  def update_api_last_access_at!
+    if self.api_access_token.present?
+      self.api_access_token[:last_access_at] = Time.now.in_time_zone(self.get_token_timezone(:api_access_token, :last_access_at))
+      self.save
+    end
+  end
+
   # return an valid access token (will renew if expired) - does not apply to api access tokens, those cannot be renewed
   def valid_access_token
     self.access_token_expired? ? self.generate_access_token : self.access_token
   end
 
   # extract timezone from an access token to allow correct date math
-  def get_token_timezone(token_method)
-    self.send(token_method)[:expires_at].zone
+  def get_token_timezone(token_method, timestamp=:expires_at)
```

Comment:
I'm not sure I'm following why you have to accept that second argument -- is it possible that the two parts of the token have different timezones? Or one of them is nil?

Reply:
It's almost always `UTC`, but I noticed in some tests that sometimes it was getting set to `EDT`, so I just wanted to make sure that whatever timestamp we're looking at, we use _that_ timestamp's zone.  I wasn't sure if this is an actual usage artifact, or just me doing something strange in a test.  But I agree - it _should_ always be `UTC` since that's what we get back from OAuth.

Reply:
This actually broke almost all the tests, so I'm taking this out - if that timestamp doesn't exist it errors out, but the `expires_at` will always be there.  And you're right - there shouldn't be a disagreement between the two, and taking this out will actually enforce that.

Reply:
872ac64